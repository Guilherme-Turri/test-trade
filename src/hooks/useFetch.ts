import React from 'react'

export const useFetchData = () => {
  const [error, setError] = React.useState<string | null>(null);
  const [loading, setLoading] = React.useState<boolean>(false);

  const request = React.useCallback(async <T> (url: string, options: RequestInit) => {
    let response;
    let json:T | null;
    setLoading(true)
    try {
      setError(null)
      response = await fetch(url, options)
      json = await response.json()
      if (!response.ok) throw new Error()
    } catch (e: any) {
      json = null;
      setError('Falha ao conectar com o servidor.')
    }
    finally {
      setLoading(false)
          }
    return { response, json } 
  }, []);
  return { error, request, loading }
} 