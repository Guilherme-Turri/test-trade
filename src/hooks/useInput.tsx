import React from 'react'

const typeField = {
  email: {
    regex: /^.{2,}$/,
    message: "Por favor, digite duas letras ou mais."
  }
}

export const useInput = () => {
  const [value, setValue] = React.useState<string>('')
  const [error, setError] = React.useState<string>('');

  const validate = (value: string) => {
    if (typeField.email.regex.test(value) === false) setError(typeField.email.message)
    else setError('');
  }

  const onChange = ({ target }: React.ChangeEvent<HTMLInputElement>) => {
    setValue(target.value)
    if (target.placeholder === "Nome") {
      validate(target.value)
    }
  }

  return (
    { value, onChange, error, setValue }
  )
}