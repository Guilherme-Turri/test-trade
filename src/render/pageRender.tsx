import { render } from '@testing-library/react';
import { ThemeProvider } from 'styled-components';
import { theme } from '../styles/theme';
import store from '../storage/store';
import { Provider } from 'react-redux';
import { BrowserRouter } from 'react-router-dom';

export const pageRender = (children: React.ReactNode) => {
  return render(
    <ThemeProvider theme={theme}>
      <Provider store={store}>
        <BrowserRouter>
          {children}
        </BrowserRouter>
      </Provider>
    </ThemeProvider>);
};