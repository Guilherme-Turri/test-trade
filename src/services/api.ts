const API_URL = ' https://v3.football.api-sports.io/'

export function AUTH_USER_API(apiKey:string) {
  return {
    url: `${API_URL}/countries`,
    options: {
      method: 'GET',
      headers: {
      'x-rapidapi-host': 'v3.football.api-sports.io',
      'x-rapidapi-key': apiKey
    },
    },
  };
}