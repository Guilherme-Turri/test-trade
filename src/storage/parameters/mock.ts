export const initialState={
  parameters:{
    country:{
      all: null,
      selected:null
    },
    league:{
      all:null, 
      selected:null
    },
    club:{
      all: null,
      selected:null
    },
    season:{
      all:null,
      selected:null,
   }
  }
  }

export const mockCountries = [
  {
    name: 'C1',
    code: '01',
    flag: 'flag1'
  }, 
  {
    name: 'C2',
    code: '02',
    flag: 'flag2'
  },
]

export const mockLeagues =[
  {
    league:{
      id:'l1',
      name:'league01',
      logo:'logo01'
    },
  },
  {
    league:{
      id:'l2',
      name:'league02',
      logo:'logo02'
    },
  }
]

export const mockClubs =[
  {
    team:{
      id:'c1',
      name:'club01',
      logo:'logo01',
      code:'123'
    },
  },
  {
    team:{
      id:'c2',
      name:'club02',
      logo:'logo02',
      code:'2345'
    },
  }
]

