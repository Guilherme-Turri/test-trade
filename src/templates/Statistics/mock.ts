export const mockPlayers = [
  {
    player:{
      name: 'Player1',
      age: 22,
      nationality: 'BR',
     }
   },
   {
    player:{
      name: 'Player2',
      age: 23,
      nationality: 'BR',
     }
   }
]



export const mockStatistcs ={

    fixtures:{
      wins: {
        total: '2'
      },
      draws: {
        total: '2'
      },
      loses: {
        total: '2'
      },
      played: {
        total: '6'
      },
    },
    goals:{
      for:{
        minute:{
          '0-15':{
            total:1
          },
          '16-30':{
            total:1
          },
          '31-45':{
            total:2
          },
          '46-60':{
            total:3
          },
          '61-75':{
            total:2
          },
          '76-90':{
            total:1
          },
          '91-105':{
            total:2
          },
          '106-120':{
            total:1
          }
        }
      }
    },
    lineups:[
      {
        formation:'4-4-2',
        played:6
      },
      {
        formation:'4-3-3',
        played:5
      },
    ],
  }




export const parameters = {
  country: {
    selected: {
      name: 'Brazil',
      flag: 'www.flag/BR'
    }
  },
  season: {
    selected: '2020'
  },
  league: {
    selected: {
      league: {
        id: '01',
        name: 'LeagueName'
      }
    }
  },
  club: {
    selected: {
      team: {
        id: '01',
        name: 'ClubName',
        logo: 'ClubLogo'
      }
    }
  }
}