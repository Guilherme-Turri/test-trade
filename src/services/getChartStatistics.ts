import { CHART_STATISTICS_API } from "./api";
import { IRequest, IPlayer } from "../types/types";

interface PlayerProps{
  response?:IPlayer[],
  errors?:{
    token:string
  }
} 

export const getChartStatistics = async (apiKey:string, league:string, season:string, club:string, request:IRequest<PlayerProps>)=> {
    const { url, options } = CHART_STATISTICS_API(apiKey, league, season, club)
  const {response, json} = await request(url, options)
    if(response && json){
    return {response, json}
   }
    else return 
 }