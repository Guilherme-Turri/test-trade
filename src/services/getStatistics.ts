import { STATISTICS_API } from "./api";
import { IRequest, IFixture, minutesT } from "../types/types";

type lineupT = {
  formation:string,
  played:number
}

interface getStatisticsProps{
  response?:{
    fixtures:IFixture,
    goals:{
      for:{
        minute:minutesT
      }
    },
    lineups:lineupT[],
  }
  errors?:{
    token:string
  }
} 


export const getStatistics = async (apiKey:string, league:string, season:string, club:string, request:IRequest<getStatisticsProps>)=> {
  
  const { url, options } = STATISTICS_API(apiKey, league, season, club)
  const {response, json} = await request(url, options)
  if(response && json){
    return {response, json}
   }
    else return 
 } 