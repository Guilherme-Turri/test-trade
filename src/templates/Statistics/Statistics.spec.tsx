/* eslint-disable testing-library/no-unnecessary-act */
import { pageRender } from "../../render/pageRender";
import { Statistics } from ".";
import { screen, act, waitFor } from '@testing-library/react'
import * as reactRedux from 'react-redux'
import * as getStatisticsTest from '../../services/getStatistics';
import * as getChartStatisticsTest from '../../services/getChartStatistics';
import * as fetchDataTest from '../../hooks/useFetch'
import { parameters, mockPlayers, mockStatistcs } from './mock'

jest.setTimeout(10000)

jest.mock('react-redux', () => ({
  ...jest.requireActual('react-redux'),
  useSelector: jest.fn(),
}));


describe('<Statistcs />', () => {
  const useSelectorMock = reactRedux.useSelector as jest.Mock;
  const getStaticsSpy = jest.spyOn(getStatisticsTest, 'getStatistics')
  const getChartStatisticsSpy = jest.spyOn(getChartStatisticsTest, 'getChartStatistics')

  it('should render on screen properly', async () => {
    useSelectorMock.mockReturnValue({
      user: {
        apiKey: '123'
      },
      modal: {
        showModal: false,
        type: ''
      },
      parameters
    });
    getStaticsSpy.mockResolvedValue({
      response: { status: 200 },
      json: {
        response: mockStatistcs,
      }
    });

    getChartStatisticsSpy.mockResolvedValue({
      response: { status: 200 },
      json: {
        response: mockPlayers,
      }
    });
    await act(async () => {
      const { container } = pageRender(<Statistics />)
      expect(container).toMatchSnapshot();
    })
  });
  it('should render on screen properly getStatits error msg', async () => {
    useSelectorMock.mockReturnValue({
      user: {
        apiKey: '123'
      },
      modal: {
        showModal: false,
        type: ''
      },
      parameters
    });
    getStaticsSpy.mockResolvedValue({
      response: { status: 200 },
      json: {
        errors: { token: 'Verifique novamente a busca' }
      }
    });

    getChartStatisticsSpy.mockResolvedValue({
      response: { status: 200 },
      json: {
        response: mockPlayers,
      }
    });

    await act(async () => {
      pageRender(<Statistics />)
    })
    await waitFor(() => {
      const msg = screen.getByText('Verifique novamente a busca')
      expect(msg).toBeInTheDocument();
    })
  });

  it('should render on screen properly getChartStatits error msg', async () => {
    useSelectorMock.mockReturnValue({
      user: {
        apiKey: '123'
      },
      modal: {
        showModal: false,
        type: ''
      },
      parameters
    });
    getStaticsSpy.mockResolvedValue({
      response: { status: 200 },
      json: {
        response: mockStatistcs
      }
    });

    getChartStatisticsSpy.mockResolvedValue({
      response: { status: 200 },
      json: {
        response: [],
        errors: { token: 'Verifique novamente a busca' }
      }
    });
    await act(async () => {
      pageRender(<Statistics />)
    })
    await waitFor(() => {
      const msg = screen.getByText('Verifique novamente a busca')
      expect(msg).toBeInTheDocument();
    })
  });
})
describe('<Statistics/> -  display right actions by the custom hook useFetch', () => {
  const useSelectorMock = reactRedux.useSelector as jest.Mock;
  it('should display loading on screen', async () => {
    jest.spyOn(fetchDataTest, "useFetchData").mockReturnValue({
      error: null,
      request: jest.fn(),
      loading: true
    });
    useSelectorMock.mockReturnValue({
      user: {
        apiKey: '123'
      },
      modal: {
        showModal: false,
        type: ''
      },
      parameters
    });

    await act(async () => {
      pageRender(<Statistics />)
    })

    await waitFor(() => {
      expect(screen.getByTestId('Ball')).toBeInTheDocument();
    })
  })

  it('should display errormsg on screen', async () => {
    jest.spyOn(fetchDataTest, "useFetchData").mockReturnValue({
      error: 'ErrorMsg',
      request: jest.fn(),
      loading: false
    });
    useSelectorMock.mockReturnValue({
      user: {
        apiKey: '123'
      },
      modal: {
        showModal: false,
        type: ''
      },
      parameters
    });
    await act(async () => {
      pageRender(<Statistics />)
    })
    await waitFor(() => {
      expect(screen.getByText('ErrorMsg')).toBeInTheDocument();
    })
  })
})   
