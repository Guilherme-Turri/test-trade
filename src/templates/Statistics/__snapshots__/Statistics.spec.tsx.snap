// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Statistcs /> should render on screen properly 1`] = `<div />`;
